File: src/_726.java
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * LeetCode 726 - Number of Atoms
 * <p>
 * LL-Parser
 * Rules:
 * 1) S -> epsilon, triggered when the next character is epsilon or ')'.
 * 2) S -> [A-Z][a-z]*[0-9]* S, triggered when the next character is [A-Z].
 * 3) S -> (S)[0-9]* S, triggered when the next character is '('.
 */
class _726 {
    int ptr = 0;
    String s;

    public String countOfAtoms(String formula) {
        ptr = 0;
        s = formula;
        Map<String, Integer> map = doit();
        return String.join("", map.keySet().stream().sorted().map(k -> k + (map.get(k) == 1 ? "" : "" + map.get(k))).collect(Collectors.toList()));
    }

    private Map<String, Integer> doit() {
        if (ptr >= s.length() || s.charAt(ptr) == ')')
            return new HashMap<>();
        else {
            Map<String, Integer> m;
            if (s.charAt(ptr) == '(') {
                ptr++; // eat '('
                m = doit();
                ptr++; // eat ')'
                int count = count();
                m.replaceAll((k, v) -> v * count);  // v * count() is wrong since this lambda will be applied to all the keys but count() can only be called once here.
            } else {
                m = new HashMap<>();
                m.put(atom(), count());
            }
            return merge(m, doit());
        }
    }

    private Map<String, Integer> merge(Map<String, Integer> m, Map<String, Integer> M) {
        if (m.size() > M.size()) return merge(M, m);
        for (String atom : m.keySet())
            M.put(atom, M.getOrDefault(atom, 0) + m.get(atom));
        return M;
    }

    private String atom() {
        String res = "" + s.charAt(ptr++);
        for (; ptr < s.length() && Character.isLowerCase(s.charAt(ptr)); ptr++)
            res += s.charAt(ptr);
        return res;
    }

    private int count() {
        String res = "";
        for (; ptr < s.length() && Character.isDigit(s.charAt(ptr)); ptr++)
            res += s.charAt(ptr);
        return res.isEmpty() ? 1 : Integer.parseInt(res);
    }

    public static void main(String[] args) {
        _726 sol = new _726();
        System.out.println(sol.countOfAtoms("Mg(OH)2"));
    }
}

File: src/_726.py
"""
LeetCode 726 - Number of Atoms

LL-Parser

Rules:
1) S -> epsilon, triggered when the next character is epsilon or ')'.
2) S -> [A-Z][a-z]*[0-9]* S, triggered when the next character is [A-Z].
3) S -> (S)[0-9]* S, triggered when the next character is '('.
"""
from collections import defaultdict


class Solution:
    def countOfAtoms(self, formula):
        """
        :type formula: str
        :rtype: str
        """

        self.ptr = 0

        def _atom(s):
            res = s[self.ptr]
            self.ptr += 1
            while self.ptr < len(s) and str.islower(s[self.ptr]):
                res += s[self.ptr]
                self.ptr += 1
            return res

        def _count(s):
            res = ""
            while self.ptr < len(s) and str.isnumeric(s[self.ptr]):
                res += s[self.ptr]
                self.ptr += 1
            return int(res) if res else 1

        def merge(m, M):
            if len(m) > len(M):
                return merge(M, m)
            for atom, cnt in m.items():
                M[atom] += cnt
            return M

        def doit(s):
            if self.ptr >= len(s) or s[self.ptr] == ')':
                return defaultdict(int)

            if s[self.ptr] == '(':
                self.ptr += 1  # eat (
                m = doit(s)
                self.ptr += 1  # eat )

                count = _count(s)
                m = defaultdict(int, {k: v * count for k, v in m.items()})
            else:
                atom = _atom(s)
                count = _count(s)
                m = defaultdict(int, {atom: count})

            return merge(m, doit(s))

        m = doit(formula)
        return ''.join('{0}{1}'.format(atom, m[atom] if m[atom] > 1 else '') for atom in sorted(m.keys()))
